### DISCUSSION TITLE
Deciding how to implement persistent storage for our database layer

### DISCUSSION THREAD

**Discussion by Poggecci** [uprm-inso4116-2023-2024-S1/semester-project-college-toolbox#122] [Q&A]
# Problem Statement
Currently, we use a PostgreSQL database run from within a docker container. This provides a nice separation layer between our backend and database (the backend configuration could be hypothetically connect to any postgres server). However, this means every developer on our team has their own database with their own data. While this ensures every developer can play with the database and work quickly, it limits their ability to test applications built on existing datasets like Existing Solutions and the Tuition Scheduler. To ensure developers have an environment most similar to production, we need a way to persist our datasets in the repository or some other resource.

Light disclaimer: I am calling this issue persistent storage as currently even though data is stored between runs on each developer's database container (so it is persistent), deleting the container wipes the data, and we don't have a great way of sharing containers (and that would go against the point of docker anyway, as volumes are what you share, not containers).

## Ideas
Here are some ideas I have thought of to tackle this problem. Please comment which one you like or suggest a different alternative. (they are roughly ordered in my preference)
### Migrate to an in-memory database (SQLite)
SQLite stores the whole database transparently (using this word to mean easy to find and with predictable behavior) in a .db file. We can store this file in the repo and add it to the .gitignore (preventing developer changes from altering the copy on the repo, but still letting developers pulling the code have access to a prepopulated starting point)

### Manual Import Script
Since our datasets are already in an easy to import format (the Tuition Scheduler data is in a SQLite db and the Existing Solutions is in an Excel spreadsheet), we can write a script that takes in the dataset and adds the data to the Postgres DB (or we just do the import process ourselves by using something like DBeaver)

### Docker Volumes
We use docker volumes to tell docker to store the contents of the database in a predetermined place in the local disk. We then add a volume with all our datasets to the repo and have developers use that

### Hosting
Have someone or some service host a production database to which we can connect to. This would be nice regardless but I don't want to spend money on it and most ISPs in Puerto Rico are stingy and do not offer port forwarding unless you pay for it.

**Comment by Juliam-Rodriguez** [comment]
You have clearly put some thought into it, so I am sure you might have already considered the following and for some limitation its not a good idea, but since hosting is probably the "ideal" case here, I know that google cloud platform and AWS provide a free hosting service with some limitations, would this be a viable option? we won't have to pay and we get the benefit of it being hosted and sharing the same database.

If this was not a possibility, I think how you described migrating to SQLite sounds the best, tho I am assuming then that all changes to the database would be affected only locally? and any real change would have to be done by a designated person that pushes the .db file to the repo? Manual Import Scripting would also be a pretty good idea if doing a bit of "extra work" is not a problem. We don't have to use any external setup with docker and we don't risk having garbage data in our database since we can just re-run the script to reset it (sqLite solution is kinda similar since we can just use version control to reset the db correct me if I'm wrong), though this would also bring the issue of having changes done only locally, but I feel like this isn't really an issue because if we really want all changes to be relevant in deployment we would still need to host it.

Lastly I think we could use docker volumes to have persistent data, I'm not entirely familiar with it but depending how willing we are as a team we could all set it up so we can have the same data and any data that must be added could just be updated in the volume (if that's how it even works lmao). The problem with this solution is I know there is a bit of a learning curve and not sure how viable using this would be for the rest of the team (myself included). If we are considering this option, I think a combination with manual scripting is a good idea, this way we have a simple way of populating the data and also a way of resetting the database easily if something goes wrong.

that being said... IMO:

Assuming hosting is not a possibility, I think the best options are either SQLite or Manual Scripting (depending on popularity). So those would have my vote.

In terms of what I think would be the best one, docker volumes with manual scripting over the others, but considering complexity, I would go for any of the previously mentioned.
